### PR TITLE
add rating filter (fix #11301)

### LINKED ISSUE #11301: Add filter for rating
### Tell us your idea!

I did not find a possibility to filter caches for specific ratings

### What solution would you suggest?

Range filter like difficulty and terrain 

### Is there an alternative?

No or I did not see it

### Additional Information

2021.07.17-RC2

### PR DESCRIPTION
**Pull request by moving-bits**
## Description
add a rating filter to the filter framework

### CODE REVIEW

File: main/src/cgeo/geocaching/connector/gc/GCMap.java [hunk 1]
```diff
@@ -187,6 +188,10 @@ private static void fillForBasicFilter(@NonNull final BaseGeocacheFilter basicFi
                 fillForBasicFilter(((DifficultyAndTerrainGeocacheFilter) basicFilter).difficultyGeocacheFilter, search);
                 fillForBasicFilter(((DifficultyAndTerrainGeocacheFilter) basicFilter).terrainGeocacheFilter, search);
                 break;
+            case RATING:
+                // not supported for online searches
+                // search.setRating(((RatingGeocacheFilter) basicFilter).getMinRangeValue(), ((RatingGeocacheFilter) basicFilter).getMaxRangeValue());
+                break;
```
Comment (eddiemuc) [inline main/src/cgeo/geocaching/connector/gc/GCMap.java:194]: Ah this is what you meant? Yes, you can leave RATING out in the gcconnector.
But add it to Okapiclient (and also to the list of filters supported by it in implementation of `ISearchByFilter`)

File: main/src/cgeo/geocaching/connector/oc/OkapiClient.java [hunk 2]
```diff
@@ -329,6 +330,12 @@ private static void fillForBasicFilter(final BaseGeocacheFilter basicFilter, fin
                 fillForBasicFilter(((DifficultyAndTerrainGeocacheFilter) basicFilter).difficultyGeocacheFilter, params, valueMap, connector);
                 fillForBasicFilter(((DifficultyAndTerrainGeocacheFilter) basicFilter).terrainGeocacheFilter, params, valueMap, connector);
                 break;
+            case RATING:
+                final RatingGeocacheFilter ratingFilter = (RatingGeocacheFilter) basicFilter;
+                if (ratingFilter.getMinRangeValue() != null) {
+                    valueMap.put("rating", (ratingFilter.getMinRangeValue() == null ? "1" : ((int) Math.floor(ratingFilter.getMinRangeValue()))) + "-" + (ratingFilter.getMaxRangeValue() == null ? "5" : Math.round(ratingFilter.getMaxRangeValue())));
+                }
+                break;
```
Comment (eddiemuc) [inline main/src/cgeo/geocaching/connector/oc/OkapiClient.java:338]: Exactly. And don't forget to make ISearchByFilter.getFilterCapabilities also return the new RATING type in its result

File: main/src/cgeo/geocaching/connector/gc/GCMap.java (hunk 1 above)
Comment (moving-bits) [reply main/src/cgeo/geocaching/connector/gc/GCMap.java:192]: Thanks, have added it

File: main/src/cgeo/geocaching/filters/gui/FilterViewHolderCreator.java [hunk 3]
```diff
@@ -168,7 +169,7 @@ public static void setListInfo(final Collection<Geocache> filteredList, final bo
         listInfoIsComplete = isComplete;
     }
 
-    private static IFilterViewHolder<?> createTerrainDifficultyFilterViewHolder() {
+    private static IFilterViewHolder<?> create1to5ItemRangeSelectorViewHolder() {
```
Comment (eddiemuc) [inline main/src/cgeo/geocaching/filters/gui/FilterViewHolderCreator.java:172]: Nice reusage idea :-)

File: main/src/cgeo/geocaching/connector/oc/OkapiClient.java (hunk 2 above)
Comment (moving-bits) [reply main/src/cgeo/geocaching/connector/oc/OkapiClient.java:338]: Have added it, thanks

File: main/src/cgeo/geocaching/filters/gui/FilterViewHolderCreator.java (hunk 3 above)
Comment (moving-bits) [reply main/src/cgeo/geocaching/filters/gui/FilterViewHolderCreator.java:172]: This was the reason why I did not even look for its own ViewHolder originally - I thought that the UI was associated implicitly by using a `NumberRangeGeocacheFilter`. But adding it was simple enough, just had to know that it's necessary.
So far my first coding experience with the new filter framework was a real pleasure - thanks for putting all this effort into it!

File: main/src/cgeo/geocaching/connector/gc/GCMap.java [hunk 4]
```diff
@@ -187,6 +187,9 @@ private static void fillForBasicFilter(@NonNull final BaseGeocacheFilter basicFi
                 fillForBasicFilter(((DifficultyAndTerrainGeocacheFilter) basicFilter).difficultyGeocacheFilter, search);
                 fillForBasicFilter(((DifficultyAndTerrainGeocacheFilter) basicFilter).terrainGeocacheFilter, search);
                 break;
+            case RATING:
```
Comment (eddiemuc) [inline main/src/cgeo/geocaching/connector/gc/GCMap.java:190]: You can simply leave out the "case RATING" statement, the "not-supported-case" is handled by "default"

**Review by eddiemuc — APPROVED** [review]
From code-side this seems to be ready for merge.
@moving-bits do you plan to work on this any further or can the merge be executed?

File: main/src/cgeo/geocaching/connector/gc/GCMap.java (hunk 4 above)
Comment (moving-bits) [reply main/src/cgeo/geocaching/connector/gc/GCMap.java:190]: Yes, I know. I left it in as explicit documentation that this connector is not able to handle such requests. But I can remove it, if requested.

**Comment by moving-bits** [comment]
Thanks for the review.

> From code-side this seems to be ready for merge.
> @moving-bits do you plan to work on this any further or can the merge be executed?

ready from my side, except if I shall remove my "documentation" in gc api.
Changed the PR type accordingly.

**Comment by eddiemuc** [comment]
> ready from my side, except if I shall remove my "documentation" in gc api.
> Changed the PR type accordingly.

I don't see this as a merge-breaker :-) So let's merge. Thanks a lot for diving into filtering!

**Pull request merged**